### PR TITLE
WIP: Add support for native JSON type support added from DB21.3 

### PR DESCRIPTION
**Pull request by sudarshan12s**
# godror JSON related changes.

**Note** ODPI changes are under review, they shall be submitted in odpic github separately. below is added just for testing

- Add support for JSON DB type which is introduced from DB 21.3 release. Earlier BLOB, CLOB , VARCHAR2 were used to stored JSON data 

- various json inputs that can be inserted or read from JSON type column  
          - JSON formatted strings. eJSON strings are not added, shall be added in next review
          - map, array , scalars (int, float, string, time.Time, time.Duration, byte[], bool)
- while reading JSON column, we use godror.JSONOptNumberAsString as option to read NUMBER types stored in DB and return godror.Number

**Table below indicates DB Native types used for the Go types:**

<table class="wrapped confluenceTable"><colgroup><col style="width: 204.0px;" /><col style="width: 163.0px;" /><col style="width: 191.0px;" /></colgroup><tbody><tr><th class="confluenceTh">SQL&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</th><th class="confluenceTh">Go&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</th></tr><tr><td class="confluenceTd"><p>NUMBER</p></td><td class="confluenceTd"><p>Int8, int16, int32, int64, int, uint, uint8, uint16, uint32, uint64, float32, float64</p></td></tr><tr><td class="confluenceTd">VARCHAR2 (string)</td><td class="confluenceTd">string</td></tr><tr><td class="confluenceTd"><span style="color: #172b4d;">TIMESTAMP</span></td><td class="confluenceTd">time.Time</td></tr><tr><td class="confluenceTd"><span style="color: #172b4d;">INTERVAL DAY TO SECOND</span></td><td class="confluenceTd">time.Duration</td></tr><tr><td class="confluenceTd">RAW</td><td class="confluenceTd"><p><span class="placeholder-inline-tasks">[]byte</span></p></td></tr><tr><td class="confluenceTd">boolean</td><td class="confluenceTd">bool</td></tr><tr><td class="confluenceTd">object</td><td class="confluenceTd">map[string]interface{}</td></tr><tr><td class="confluenceTd">array</td><td class="confluenceTd"><ul class="inline-task-list" data-inline-tasks-content-id="3053177290"><li data-inline-task-id="1"><span class="placeholder-inline-tasks">[]interface{}</span></li></ul></td></tr></tbody></table>

**Table below indicates converted Go types after reading native types from DB:**
<table class="confluenceTable wrapped"><colgroup><col /><col /><col /></colgroup><tbody><tr><th class="confluenceTh">SQL</th><th class="confluenceTh">Go</th></tr><tr><td class="confluenceTd"><p>NUMBER</p></td><td class="confluenceTd"><p>godror.Number or float64</p></td></tr><tr><td class="confluenceTd">boolean</td><td class="confluenceTd">bool</td></tr><tr><td class="confluenceTd">object</td><td class="confluenceTd"><p>map[string]interface{}&nbsp;</p><p><br /></p><p><br /></p><p><br /></p><p><br /></p><p><br /></p><p><br /></p></td></tr><tr><td class="confluenceTd">array</td><td class="confluenceTd"><ul class="inline-task-list" data-inline-tasks-content-id="3053177290"><li data-inline-task-id="2"><span class="placeholder-inline-tasks">[]interface{}</span></li></ul></td></tr><tr><td class="confluenceTd">RAW</td><td class="confluenceTd">byte[]</td></tr><tr><td class="confluenceTd"><span style="color: #172b4d;">INTERVAL DAY TO SECOND</span></td><td class="confluenceTd">time.Duration</td></tr><tr><td class="confluenceTd">VARCHAR2(string)</td><td class="confluenceTd">string</td></tr></tbody></table>

**Note:** All int8, int16, int32, int64, float32, float64, uint8, uint16, uint32, uint64 are stored as NUMBER in DB. When they are retrieved they are returned as double (if option, godror.JSONOptDefault is selected) else they are returned as godror.Number type. 

### JSONValue: 
JSONValue is the interface implemented by structs encapsulating various JSON inputs.
// JSON, JSONObject, JSONArray and JSONScalar implement it.

### JSONString:
type representing JSON/eJSON formatted strings.

### JSONScalar:
type representing scalar values such as int, float, bool, time.Time, time.Duration, byte[], map and arrays.

### JSON:
type which represents root node in JSON tree

### JSONObject:
type which represents scalar type Object

### JSONArray:
type which represents scalar type Array

## Example Usage:

**Table assumed is :**
```
CREATE TABLE test (id NUMBER(6), jdoc JSON)
```

### Inserting/Reading JSON Formatted string: 
Applications(REST service) dealing with JSON strings . 

#### Insert:
```
jsonstring := "{\"person\":{\"BirthDate\":\"1999-02-03T00:00:00\",\"ID\":\"12\",\"JoinDate\":\"2020-11-24T12:34:56.123000Z\",\"Name\":\"Alex\",\"RandomString\":\"APKZYKSv2\",\"age\":\"25\",\"creditScore\":[\"700\",\"250\",\"340\"],\"salary\":\"45.23\"}}"

jsonval := godror.JSONString{Value: jsonstring, Flags: 0}
stmt.ExecContext(ctx, id, jsonval)
//Flags indicate if the JSON String has extended types which can be mentioned through eJSON format. 
```

#### Read:
```
var id  interface{}
var jsondoc godror.JSON
rows.Scan(&id, &jsondoc)
fmt.Println(jsondoc)
```

###  Inserting/Reading Go map :
Applications dealing with Go maps handling updates before pushing to DB
 
#### Insert:
```
jsmap := map[string]interface{}{
        "person": map[string]interface{}{
            "ID":        godror.Number("12"),
            "FirstName": "Mary",
            "LastName":  "John",
            "creditScore": []interface{}{
                godror.Number("123456789123456789123456789123456789.12"),
                godror.Number("250"),
                godror.Number("340"),
            },
            "age":              godror.Number("25"),
            "BirthDate":        birthdate,
            "salary":           godror.Number("45.23"),
            "Local":            true,
            "BinData":          []byte{0, 1, 2, 3, 4},
            "TravelTimePerDay": travelTime,
        },
    }
jsonval, _ := godror.JSONValue{Value:jsmap}
stmt.ExecContext(ctx, id, jsonval)

```

#### Read:
```
var id interface{}
var jsondoc godror.JSON 
rows.Scan(&id, &jsondoc)

// Get map[string]interface{}
var gotmap map[string]interface{}
v, err = jsondoc.GetValue(godror.JSONOptNumberAsString)
gotmap, ok = v.(map[string]interface{});
```
 

### Inserting/Reading Array:
Applications dealing with Go Arrays handling updates before pushing to DB

#### Insert:
```
 jsarr := []interface{}{
        map[string]interface{}{
            "person": map[string]interface{}{
                "ID":        godror.Number("12"),
                "FirstName": "Mary",
                "LastName":  "John",
                "creditScore": []interface{}{
                    godror.Number("700"),
                    godror.Number("250"),
                    godror.Number("340"),
                },
                "age":       godror.Number("25"),
                "BirthDate": birthdate,
                "salary":    godror.Number("4500.2351"),
                "Local":     true,
            },
        },
}

jsonval, _ := godror.JsonValue{Value:jsarr}
stmt.ExecContext(ctx, id, jsonval)

```

#### Read:
```
var id interface{}
var jsondoc godror.JSON
rows.Scan(&id, &jsondoc)

// Get JSONArray []interface{}
var gotarr []interface{}
v, err := jsondoc.GetValue(godror.JSONOptNumberAsString)
gotarr, ok = v.([]interface{}); 
```

### Reading scalar values 
```
qry := SELECT id, c.jdoc.person.BirthDate from table c
rows, _ := conn.QueryContext(ctx,qry)
var id interface{}
var personBirthDate godror.JSON
// read scalar value, BIrthDate (time.Time) 
rows.Scan(&id, &personBirthDate)

var gotDOB time.Time
birthDateScalar, err :=personBirthDate.GetValue(godror.JSONOptDefault)
gotDOB, ok = birthDateScalar.(time.Time);
```
### updating scalar values 
```
// read object , person from JSON coloumn
qry := SELECT id, c.jdoc.person FROM test  c
var v interface{}
var personMap map[string]interface{}
var jsonDoc godror.JSON
rows.Scan(&id, &jsonDoc); 
 v, err jsonDoc.GetValue(godror.JSONOptNumberAsString)
 personMap, ok = v.(map[string]interface{});

// Update BirthDate and bind the new map
personMap["BirthDate"] = newBirthDate
jsonval, err := godror.JSONValue{Value:personMap}
updateqry := "update  test  c set c.jdoc=JSON_TRANSFORM(c.jdoc, set '$.person'=:1)"
conn.ExecContext(ctx, updateqry, jsonval)

```

- Testing:
following test cases are added:

TestReadWriteJsonArray // inserts array, reads it and does reflect.DeepEqual for comparision of user array and fetched array
TestReadWriteJsonMap // does batch insert and also single insert of map.  It then  reads each row from DB and does reflect.DeepEqual for comparision of user given map and fetched object from DB
TestReadWriteJSONString // batch insert json string and reads as string and does compare 
TestReadJSONScalar // It fetches the JSON coloumn which is an array. Displays birthdates of each entry, person. It validates the birthdate matches with what is inserted.
TestUpdateJSONScalar // It retrieves the object , person from JSON column and binds scalar map and also binds LastName as string .

- 21.3 inst client and DB
https://www.oracle.com/database/technologies/instant-client/linux-x86-64-downloads.html
https://www.oracle.com/database/technologies/oracle21c-linux-downloads.html

### CODE REVIEW

**Review by tgulacsi — COMMENTED** [review]
go fmt your code, don't introduce unnecessary changes!

File: stmt.go [hunk 1]
```diff
@@ -2320,6 +2352,37 @@ func (c *conn) dataSetLOB(dv *C.dpiVar, data []C.dpiData, vv interface{}) error
...
+	i := 0
+	if len(data) == 0 {
+		return nil
+	}
+	if vv == nil {
+		return dataSetNull(dv, data, nil)
+	}
+	switch js := vv.(type) {
+	case JSONString:
+		cstr := C.CString(js.Value)
```
Comment (tgulacsi) [inline stmt.go:2366]: Who will C.free cstr ?

File: orahlp.go [hunk 2]
```diff
@@ -23,6 +23,12 @@ import (
 // Number as string
 type Number string
 
+// Timestamp as time.Time. For eJSON Marshal Support
+type Timestamp time.Time
```
Comment (tgulacsi) [inline orahlp.go:27]: Where is it used?

File: conn.go [hunk 3]
```diff
@@ -738,7 +738,7 @@ func IsBadConn(err error) bool {
 		}
 
 	case // cases by experience:
-		3106,  // fatal two-task communication protocol error
+        3106,  // fatal two-task communication protocol error
```
Comment (tgulacsi) [inline conn.go:741]: Pkease don't introduce unnecessary whitespace changes!

**Review by tgulacsi — CHANGES_REQUESTED** [review]

File: json.go [hunk 4]
```diff
@@ -0,0 +1,809 @@
...
+		}
+		C.godror_dpiJson_setBool(jsonnode, &(data.dpiData))
+	default:
+		return fmt.Errorf("Unsupported type %T\n", in)
+	}
+	return nil
+}
+
+// Creates a JSONInt from user input []interface{}
+func newJSONScalar(val interface{}, jscalar *JSONScalar) error {
```
Comment (tgulacsi) [inline json.go:648]: Why the jscalar pointer? Why not just return *JSONScalar?

File: json.go [hunk 5]
```diff
@@ -0,0 +1,809 @@
...
+
+// Frees the C memory associated with JSONObject
+func (jsobj *JSONObject) Close() error {
+	C.godror_dpiJsonfreeMem(jsobj.dpiJsonNode)
+	return nil
+}
+
+func (j JSONObject) Len() int { return int(j.dpiJsonObject.numFields) }
+func (j JSONObject) Get(m map[string]Data) {
+	n := int(j.dpiJsonObject.numFields)
```
Comment (tgulacsi) [inline json.go:700]: Why is the map an argument? Why not just return it?

File: json.go [hunk 6]
```diff
@@ -0,0 +1,809 @@
...
+			m[C.GoStringN(names[i], C.int(nameLengths[i]))] = ua
+		} else if fields[i].oracleTypeNum == C.DPI_ORACLE_TYPE_VARCHAR {
+			keyval, err := getJSONScalarString(d)
+			if err == nil {
+				m[C.GoStringN(names[i], C.int(nameLengths[i]))] = keyval
+			} else {
+				return nil, err
+			}
+		} else if fields[i].oracleTypeNum == C.DPI_ORACLE_TYPE_NUMBER {
+			jd, err := getJSONScalarDouble(d)
```
Comment (tgulacsi) [inline json.go:743]: Why double? Why not godror.Number or string?
Why do we want to lose precision?

File: json.go [hunk 7]
```diff
@@ -0,0 +1,809 @@
...
+// It is like a root node in JSON tree.
+type JSON struct {
+	dpiJson       *C.dpiJson
+	isMemOwned    bool // true indicates associated C memory needs to be freed.
+	oracleTypeNum C.dpiOracleTypeNum
+}
+
+// During Fetch the JSON struct is created.
+// Explicit creation is not supported.
+func createJSON(js *C.dpiJson, mutable bool) (JSON, error) {
```
Comment (tgulacsi) [inline json.go:307]: Why the error return when it never returns error?

**Review by tgulacsi — CHANGES_REQUESTED** [review]

File: stmt.go (hunk 1 above)
Comment (sudarshan12s) [reply stmt.go:2366]: FIxed and updated.

File: orahlp.go (hunk 2 above)
Comment (sudarshan12s) [reply orahlp.go]: It was something added for eJSON , not needed now, removed it.

File: conn.go (hunk 3 above)
Comment (sudarshan12s) [reply conn.go]: mistake, I didnt want to add this file. Fixed

File: json.go (hunk 4 above)
Comment (sudarshan12s) [reply json.go]: I didnt get it completely. You mean something like below, I submitted change in this way. Please suggest:
return &JSONScalar{..

File: json.go (hunk 5 above)
Comment (sudarshan12s) [reply json.go:701]: made it as return, took changes from your private json branch.

File: json.go (hunk 7 above)
Comment (sudarshan12s) [reply json.go]: Fixed it.

File: json.go [hunk 8]
```diff
@@ -0,0 +1,797 @@
...
+		}
+	}
+	return
+}
+
+func (j JSONScalar) getOracleJSONType() C.dpiOracleTypeNum {
+	return j.dpiJsonNode.oracleTypeNum
+}
+
+func (js *JSONScalar) Close() error {
```
Comment (tgulacsi) [inline json.go:290]: All members of a struct should use the same reference: either all should be defined on a pointer, or neither.

Current code does not need js to be a pointer,
but if we want Close to be idempotent (don't SIGSEGV on double call),
then we must nil js.dpiJsonNode like
```
dpiJsonNode := js.dpiJsonNode
js.dpiJsonNode = nil
if dpiJsonNode != nil && js.isMemOwner {
  C.godror_dpiJsonfreeMem(dpiJsonNode)
}
return nil
```

The same applies to the other types (JSONArray, JSONValue...)

File: json.go [hunk 9]
```diff
@@ -0,0 +1,797 @@
...
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return ErrInvalidJSON
+	}
+	jsonNodeToData(data, node)
+	return nil
+}
+
+// Returns JSONObject from JSON
+func (j JSON) GetJSONObject(jsobj *JSONObject, opts JSONOption) error {
```
Comment (tgulacsi) [inline json.go:327]: Why not just return a *JSONObject ?

File: json.go [hunk 10]
```diff
@@ -0,0 +1,797 @@
...
+	}
+
+	*jsobj = JSONObject{dpiJsonNode: node,
+		dpiJsonObject: C.dpiData_getJsonObject(&(d.dpiData)),
+		isMemOwned:    j.isMemOwned}
+	return nil
+}
+
+// Returns JSONArray from JSON
+func (j JSON) GetJSONArray(jsarr *JSONArray, opts JSONOption) error {
```
Comment (tgulacsi) [inline json.go:345]: Why not just  return the *JSONArray ?

File: json.go [hunk 11]
```diff
@@ -0,0 +1,797 @@
...
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return ErrInvalidJSON
+	}
+	*js = JSONScalar{dpiJsonNode: node,
+		isMemOwned: j.isMemOwned}
+	return nil
+}
+
+// Returns a Go type Value from JSON
+func (j JSON) GetValue(opts JSONOption, val *interface{}) (err error) {
```
Comment (tgulacsi) [inline json.go:373]: Why not return the `(val interface{}, err error)` ?

File: json.go [hunk 12]
```diff
@@ -0,0 +1,797 @@
...
+	elts := ((*[maxArraySize]C.dpiJsonNode)(unsafe.Pointer(j.dpiJsonArray.elements)))[:n:n]
+	for i := 0; i < n; i++ {
+		var d Data
+		jsonNodeToData(&d, &elts[i])
+		nodes = append(nodes, d)
+	}
+	return nodes
+}
+
+func getJSONScalarDouble(d Data) (val float64, err error) {
```
Comment (tgulacsi) [inline json.go:463]: Where do we really need this? We should convert (lose precision) Oracle NUMBER to float64 only as a last resort.

File: json.go [hunk 13]
```diff
@@ -0,0 +1,797 @@
...
+
+func (j JSONObject) getOracleJSONType() C.dpiOracleTypeNum {
+	return C.DPI_ORACLE_TYPE_JSON_OBJECT
+}
+
+// populates dpiJsonNode from user inputs.
+// It creates a seperate memory for the new output value, jsonnode.
+// memory from user input, in is not shared with jsonnode.
+// Caller has to explicitly free using godror_dpiJsonfreeMem
+func populateJsonNode(in interface{}, jsonnode *C.dpiJsonNode) error {
```
Comment (tgulacsi) [inline json.go:547]: Make the first argument be the destination: `jsonnode *C.dpiJsonNode, in interface{}`.

File: json.go [hunk 14]
```diff
@@ -0,0 +1,797 @@
...
+		C.godror_dpiJsonfreeMem(dpijsonnode)
+		return nil, err
+	}
+	dpidataw := new(Data)
+	jsonNodeToData(dpidataw, dpijsonnode)
+	return &JSONScalar{dpiJsonNode: dpijsonnode, isMemOwned: true}, nil
+}
+
+// Creates a JSONArray from user input []interface{}
+func newJSONArray(arr []interface{}, jsarr *JSONArray) error {
```
Comment (tgulacsi) [inline json.go:662]: Why not just return the *JSONArray ?

File: json.go [hunk 15]
```diff
@@ -0,0 +1,797 @@
...
+		return err
+	}
+	dpidataw := new(Data)
+	jsonNodeToData(dpidataw, dpijsonnode)
+	*jsobj = JSONObject{dpiJsonObject: C.dpiData_getJsonObject(&(dpidataw.dpiData)), dpiJsonNode: dpijsonnode, isMemOwned: true}
+	return nil
+}
+
+// Frees the C memory associated with JSONObject
+func (jsobj *JSONObject) Close() error {
```
Comment (tgulacsi) [inline json.go:692]: Make this idempotent.

File: json.go [hunk 16]
```diff
@@ -0,0 +1,797 @@
...
+			} else {
+				nodes = append(nodes, d.Get())
+			}
+		}
+	}
+	return nodes, nil
+}
+
+// Frees the C memory allocated
+func (jsarr *JSONArray) Close() error {
```
Comment (tgulacsi) [inline json.go:525]: Make this idempotent.

File: json.go [hunk 17]
```diff
@@ -0,0 +1,797 @@
...
+		return ErrInvalidType
+	}
+	*jsarr = JSONArray{dpiJsonNode: node,
+		dpiJsonArray: C.dpiData_getJsonArray(&(d.dpiData)),
+		isMemOwned:   j.isMemOwned}
+	return nil
+}
+
+// Returns JSONScalar from JSON
+func (j JSON) GetJSONScalar(js *JSONScalar, opts JSONOption) error {
```
Comment (tgulacsi) [inline json.go:362]: Why not just return the *JSONScalar?

**Review by tgulacsi — CHANGES_REQUESTED** [review]

File: json.go (hunk 6 above)
Comment (sudarshan12s) [reply json.go:746]: working on this to retain Number precision

File: json.go (hunk 11 above)
Comment (sudarshan12s) [reply json.go:373]: I was trying to make it look similar to json Unmarshal. is it fine? I can change it ...
func Unmarshal(data []byte, v interface{}) error

Comment (sudarshan12s) [reply json.go:373]: Let me check, you mean to remain consistent with other type; JSONObject, . Let me validate on these.

File: json.go (hunk 6 above)
Comment (sudarshan12s) [reply json.go]: used getJSONScalarNumber to return godror.Number for option, JSONOptNumberAsString and float64 for JSONOptDefault . JSONOptDefault returns double and hence it is converted to float64 and probably this is used, if user is aware of the precision .

File: json.go (hunk 12 above)
Comment (sudarshan12s) [reply json.go]: Please see getJSONScalarNumber function, if it looks ok?

File: json.go (hunk 15 above)
Comment (sudarshan12s) [reply json.go:710]: Fixed

File: json.go (hunk 16 above)
Comment (sudarshan12s) [reply json.go:540]: Fixed

File: json.go (hunk 11 above)
Comment (tgulacsi) [reply json.go:392]: Unmarshal uses the destination interface{} (NOT *interface{} !) - which must contain a pointer -
to deduce the wanted type of data. It could Unmarshal a js number (float64) into a string, int, godror.Number etc.

Here GetValue does nothing like, it decides the returned data type by opts.

**Review by tgulacsi — CHANGES_REQUESTED** [review]
Following what Unmarshal does may help the users,
but may make the library code more complex.

Providing an option for GetValue complicates the using,
but simplifies the library and makes all conversions explicit.

Designing an API is not an easy task.

First I'd implement the bare minimum that allows to get/set data from/to JSON
(even by explicitly using godror.Data), and when it took shape, and all the finicky details (where to use pointer semantics, for example) are fixed, THEN start thinking about a nicer API. 
At that time maybe some helper functions would suffice - or not, and some thin wrapping layer will be needed, I don't know yet.

So I'd split this up to thinner layers.

File: json.go (hunk 4 above)
Comment (sudarshan12s) [reply json.go]: Fixed.

File: json.go (hunk 8 above)
Comment (sudarshan12s) [reply json.go:308]: Fixed and updated

File: json.go (hunk 9 above)
Comment (sudarshan12s) [reply json.go]: Fixed

File: json.go (hunk 10 above)
Comment (sudarshan12s) [reply json.go]: Fixed

File: json.go (hunk 11 above)
Comment (sudarshan12s) [reply json.go]: Made to return pointers in all the API.

File: json.go (hunk 13 above)
Comment (sudarshan12s) [reply json.go]: Fixed

File: json.go (hunk 17 above)
Comment (sudarshan12s) [reply json.go]: Fixed

**Comment by sudarshan12s** [comment]
> Following what Unmarshal does may help the users,
> but may make the library code more complex.
> 
> Providing an option for GetValue complicates the using,
> but simplifies the library and makes all conversions explicit.
> 
> Designing an API is not an easy task.
> 
> First I'd implement the bare minimum that allows to get/set data from/to JSON
> (even by explicitly using godror.Data), and when it took shape, and all the finicky details (where to use pointer semantics, for example) are fixed, THEN start thinking about a nicer API.
> At that time maybe some helper functions would suffice - or not, and some thin wrapping layer will be needed, I don't know yet.
> 
> So I'd split this up to thinner layers.

Yes Agreed. I was thinking to enhance this to include other return values  (*Data, map[string]Data, []Data, .. )  in later stages and see if i can keep API same,  but it was not trivial. 
Yes once you validate the current ones, i can contribute for the new changes.

File: json.go [hunk 18]
```diff
@@ -0,0 +1,687 @@
...
+		} else if fields[i].oracleTypeNum == C.DPI_ORACLE_TYPE_NUMBER {
+			m[C.GoStringN(names[i], C.int(nameLengths[i]))] = getJSONScalarNumber(d)
+		} else {
+			m[C.GoStringN(names[i], C.int(nameLengths[i]))] = d.Get()
+		}
+	}
+	return m, nil
+}
+
+// It populates the fields of struct taking pointer to struct.
```
Comment (tgulacsi) [inline json.go:552]: "GetInto populates the fields ..."
Start the method's comment with the method's name.

File: json.go [hunk 19]
```diff
@@ -0,0 +1,687 @@
...
+	return &JSONArray{dpiJsonArray: C.dpiData_getJsonArray(&(d.dpiData))}, nil
+}
+
+// Returns JSONScalar from JSON
+func (j JSON) GetJSONScalar(opts JSONOption) (*JSONScalar, error) {
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return nil, ErrInvalidJSON
+	}
+	return &JSONScalar{dpiJsonNode: node}, nil
```
Comment (tgulacsi) [inline json.go:294]: JSONScalar does not need to be a pointer (it has no pointer methods).
Maybe a `JSONScalar{dpiJsonNode:node}` would be better?

File: json.go [hunk 20]
```diff
@@ -0,0 +1,687 @@
...
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return nil, ErrInvalidJSON
+	}
+	var d Data
+	jsonNodeToData(&d, node)
+	if C.dpiOracleTypeNum(node.oracleTypeNum) != C.DPI_ORACLE_TYPE_JSON_ARRAY {
+		return nil, ErrInvalidType
+	}
+	return &JSONArray{dpiJsonArray: C.dpiData_getJsonArray(&(d.dpiData))}, nil
```
Comment (tgulacsi) [inline json.go:285]: JSONArray does not need to be a pointer (it has no pointer methods).
Maybe a JSONArray{...} would be better?

File: json.go [hunk 21]
```diff
@@ -0,0 +1,687 @@
...
+	var node *C.dpiJsonNode
+	var d Data
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return nil, ErrInvalidJSON
+	}
+	jsonNodeToData(&d, node)
+	if C.dpiOracleTypeNum(node.oracleTypeNum) != C.DPI_ORACLE_TYPE_JSON_OBJECT {
+		return nil, ErrInvalidType
+	}
+	return &JSONObject{dpiJsonObject: C.dpiData_getJsonObject(&(d.dpiData))}, nil
```
Comment (tgulacsi) [inline json.go:271]: JSONOBject does not need to be a pointer (it has no pointer methods).
Maybe a JSONObject{...} would be better?

**Review by tgulacsi — CHANGES_REQUESTED** [review]

File: json.go (hunk 21 above)
Comment (sudarshan12s) [reply json.go]: changed it now. Yes I think we may not add more members to JSONObject which otherwise would cause member copies if pointer receiver was not used. For now, I dont see to add more members other than the dpiJsonObject pointer.

File: json.go (hunk 20 above)
Comment (sudarshan12s) [reply json.go]: changed it now. Yes I think we may not add more members to JSONArray which otherwise would cause member copies if pointer receiver was not used. For now, I dont see to add more members other than the dpiJsonArray pointer.

File: json.go (hunk 19 above)
Comment (sudarshan12s) [reply json.go]: changed it now. Yes I think we may not add more members to JSONScalar which otherwise would cause member copies if pointer receiver was not used. For now, I dont see to add more members other than the dpiJsonNode pointer.

File: json.go (hunk 18 above)
Comment (sudarshan12s) [reply json.go]: used the function names. Followed the same convention for documenting other functions too.

File: json.go [hunk 22]
```diff
@@ -0,0 +1,733 @@
...
+
+var ErrInvalidJSON = errors.New("Invalid JSON Document")
+var ErrInvalidType = errors.New("Invalid JSON Scalar Type")
+
+const (
+	JSONOptDefault        = JSONOption(C.DPI_JSON_OPT_DEFAULT)
+	JSONOptNumberAsString = JSONOption(C.DPI_JSON_OPT_NUMBER_AS_STRING)
+)
+
+var NullJSONObject = JSONObject{}
```
Comment (tgulacsi) [inline json.go:250]: Why are these needed at all? "NullJSONObject" is longer than "JSONObject{}"!

But at least, unexport (start with lower case letter) them!

As it is know, anybody can accidentally modify them, and all functions that return a NullJSONObject,
may return something else!

File: json.go [hunk 23]
```diff
@@ -0,0 +1,733 @@
...
+	return nil
+}
+
+// GetJSONObject retrieves JSONObject from JSON based on option, opts.
+// It returns error if JSON doesnt represent object type.
+func (j JSON) GetJSONObject(opts JSONOption) (JSONObject, error) {
+	var node *C.dpiJsonNode
+	var d Data
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return NullJSONObject, ErrInvalidJSON
```
Comment (tgulacsi) [inline json.go:276]: JSONObject{} or nullJSONObject

File: json.go [hunk 24]
```diff
@@ -0,0 +1,733 @@
...
+	}
+	return JSONObject{dpiJsonObject: C.dpiData_getJsonObject(&(d.dpiData))}, nil
+}
+
+// GetJSONArray retrieves JSONArray from JSON based on option, opts.
+// It returns error if JSON doesnt represent array type.
+func (j JSON) GetJSONArray(opts JSONOption) (JSONArray, error) {
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return NullJSONArray, ErrInvalidJSON
```
Comment (tgulacsi) [inline json.go:290]: JSONArray{} or nullJSONArray

File: json.go [hunk 25]
```diff
@@ -0,0 +1,733 @@
...
+		return NullJSONArray, ErrInvalidType
+	}
+	return JSONArray{dpiJsonArray: C.dpiData_getJsonArray(&(d.dpiData))}, nil
+}
+
+// GetJSONScalar retrieves JSONScalar from JSON based on option, opts.
+func (j JSON) GetJSONScalar(opts JSONOption) (JSONScalar, error) {
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return NullJSONScalar, ErrInvalidJSON
```
Comment (tgulacsi) [inline json.go:304]: JSONScalar{} or nullJSONScalar

**Review by tgulacsi — CHANGES_REQUESTED** [review]

File: json.go (hunk 22 above)
Comment (sudarshan12s) [reply json.go]: Removed the exported symbols. returned respective structs.

File: json.go (hunk 23 above)
Comment (sudarshan12s) [reply json.go]: used JSONObject{}

File: json.go (hunk 24 above)
Comment (sudarshan12s) [reply json.go]: used JSONArray{}

File: json.go (hunk 25 above)
Comment (sudarshan12s) [reply json.go]: used JSONScalar{}

**Comment by tgulacsi** [comment]
LGTM.

So, now we're waiting for a new ODPI-C release?

**Comment by sudarshan12s** [comment]
> LGTM.
> 
> So, now we're waiting for a new ODPI-C release?

odpi-c changes are already in github.
https://github.com/oracle/odpi/commit/be1096c790fd0969d50da33ce25e5577883b9376

**Comment by tgulacsi** [comment]
Do you want some other changes now that this big has been merged, or shall I tag it as v0.26.0 ?

**Comment by sudarshan12s** [comment]
> Do you want some other changes now that this big has been merged, or shall I tag it as v0.26.0 ?

Thanks. I see tag has been created. There are no immediate changes planned now.

**Comment by cjbj** [comment]
Thanks @tgulacsi , @sudarshan12s, and team for landing this PR!
